### PR TITLE
Feature: Solution

### PR DESCRIPTION
**Pull request by antonzaharia**
## Added smaller task to user stories (Update the readme as well but added them here for more clarity):
- [x] The frontend is not properly fetching data, i.e. filters aren't being sent, requests are being sent too often, non-successful response codes aren't being handled, etc.
  - [x] Ensured filters are sent with requests
  - [x] Debouncer: Control the frequency of API requests, ensuring that they're made only after a user has paused their input, reducing unnecessary server load.
  - [x] Added error handling for API responses, displaying user-friendly messages and preventing the application from crashing on non-successful response codes.
- [x] The backend is taking too long to return the expected data, i.e.  filters aren't being applied, too much data is being returned, db performance isn't optimized, etc.
  - [x] Pagination: To reduce load times by limiting the amount of data sent in a single request.
  - [x] Filterer Concern: To cleanly apply filters, improving query efficiency and code maintainability.
  - [x] Serializers: To only send necessary data, reducing bandwidth and speeding up responses.
  - [x] Database Optimization: To enhance query speed and overall application performance.
- [x] The confidence in our current functionality and ability to deploy additional changes is extremely low.  Help us increase our confidence in an automated way.
  - [x] Unit Tests: Cover individual components or methods, ensuring they work as expected in isolation.
  - [x] Controller Tests: Verify that the application's HTTP endpoints respond as intended, covering various scenarios including success, failure, and edge cases.
  - [x] System Tests: Implement systems tests to ensure different parts of the application work together as expected.
- [x] The confidence in our security posture is also very low.  Help us increase our confidence that we are running a securely designed application.
  - [x] Strong Parameters: Use strong parameters to prevent mass-assignment vulnerabilities and ensure we explicitly permit only the required parameters for each action.
  - [x] Input Validation: Ensure all inputs from users are validated both in terms of type and content to prevent SQL injection and other forms of injection attacks.
  - [x] Update Dependencies: Use `bundler-audit` to check for vulnerabilities.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -4,7 +4,7 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 ruby "3.2.2"
 
 # Bundle edge Rails instead: gem "rails", github: "rails/rails", branch: "main"
-gem "rails", "~> 7.0.8"
+gem "rails", "~> 7.1.1"
```
Comment (antonzaharia) [inline Gemfile:7]: Updated to fix bundle audit vulnerabilities

File: Gemfile [hunk 2]
```diff
@@ -51,10 +51,24 @@ gem "bootsnap", require: false
...
+gem "active_model_serializers", "~> 0.10.14"
+
+# Pagination
+gem "will_paginate", "~> 4.0"
+
+group :test do
+  gem "minitest", "~> 5.22"
+  gem "capybara", "~> 3.40"
+  gem "selenium-webdriver", "~> 4.18"
+end
```
Comment (antonzaharia) [inline Gemfile:64]: Gems for testing

File: Gemfile [hunk 3]
```diff
@@ -51,10 +51,24 @@ gem "bootsnap", require: false
...
+
 group :development, :test do
   # See https://guides.rubyonrails.org/debugging_rails_applications.html#debugging-with-the-debug-gem
   gem "debug", platforms: %i[ mri mingw x64_mingw ]
   gem "faker"
+  gem "pry", "~> 0.14.2"
```
Comment (antonzaharia) [inline Gemfile:70]: My favourite debugging gem

File: app/javascript/components/Home.jsx [hunk 4]
```diff
@@ -1,73 +1,96 @@
...
+  const [minEmployee, setMinEmployee] = useState('')
+  const [minimumDealAmount, setMinimumDealAmount] = useState('')
+
+  const handlePageChange = (page) => {
+    setCurrentPage(page)
+    fetchCompanies(page)
+  }
+
+  const queryParams = new URLSearchParams({ companyName, industry, minEmployee, minimumDealAmount }).toString()
+  const companiesUrl = `/api/v1/companies?${queryParams}`
```
Comment (antonzaharia) [inline app/javascript/components/Home.jsx:36]: I feel that here are too many state variables and we could improve that further but for the time being I left them like that

File: app/javascript/components/Home.jsx [hunk 5]
```diff
@@ -1,73 +1,96 @@
...
+        }
+        setCompanies(data.companies)
+        setTotalPages(data.meta.pages)
+        setCurrentPage(data.meta.page)
+        setIsInitialMount(false)
+      })
+      .catch((error) => setError(`Fetch error: ${error}`))
+  }
+
+  // Custom hook to debounce feting the companies
```
Comment (antonzaharia) [inline app/javascript/components/Home.jsx:60]: fetching* - typo

File: app/javascript/components/shared/Error.jsx [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import React from 'react'
+
+function Error({ message }) {
+  return (
+    <div className="alert alert-warning" role="alert">
+      {message}
+    </div>
+  )
+}
+export default Error
```
Comment (antonzaharia) [inline app/javascript/components/shared/Error.jsx:10]: This was created with the idea of having multiple types of alerts. Probably it was supposed to be called Warning/Alert. However, I didn't need any other type, so I just left it like that for now.

File: app/models/concerns/filterer.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    # Filtering by min deal amount
+    def filter_by_minimum_deal_amount(results, params)
+      return results unless params[:minimumDealAmount].present?
+
+      ids = Deal.where('amount >= ?', params[:minimumDealAmount]).distinct.pluck(:company_id)
+      results.where(id: ids)
+    end
+  end
+end
```
Comment (antonzaharia) [inline app/models/concerns/filterer.rb:53]: In the last commit, I refactored this concern, implementing the chain of responsibility design pattern. (And it was very useful to have tests)

File: app/models/concerns/filterer.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      results.where("employee_count >= ?", params[:minEmployee])
+    end
+
+    # Filtering by min deal amount
+    def filter_by_minimum_deal_amount(results, params)
+      return results unless params[:minimumDealAmount].present?
+
+      ids = Deal.where('amount >= ?', params[:minimumDealAmount]).distinct.pluck(:company_id)
+      results.where(id: ids)
+    end
```
Comment (antonzaharia) [inline app/models/concerns/filterer.rb]: I was a bit unsure of what this filter should do exactly because filtering by the minimum deal amount is something that we cannot verify visually.
I was tempted to implement the filter by the minimum of the total deal amount because that is what we see in the last column of the table and not any actual deal amount.

However, I decided to filter by the minimum deal amount to make sure I respected the naming and followed the requirements.

**Review by antonzaharia — COMMENTED** [review]
Thanks for the clarification!

I have updated the filter accordingly and I also found two problems:
1. Double request on page load
2. N+1 Query for deal amounts sum

Both are fixed now 👍
